felix/labelindex/labelnamevalueindex/label_name_value_index.go

```diff
@@ -0,0 +1,347 @@
+	nameOfTrackedItems    string
+	allValues             map[ItemID]Item
+	labelNameToValueToIDs map[string]values[ItemID]
+}
+
+type Labeled interface {
+	OwnLabels() map[string]string
+}
+
+func New[ItemID comparable, Item Labeled](nameOfTrackedItems string) *LabelNameValueIndex[ItemID, Item] {
+	return &LabelNameValueIndex[ItemID, Item]{
+		nameOfTrackedItems:    nameOfTrackedItems,
+		allValues:             map[ItemID]Item{},
+		labelNameToValueToIDs: map[string]values[ItemID]{},
+	}
+}
+
+type values[T comparable] struct {
+	m     map[string]set.Set[T]
+	count int
+}
+
+// Add an item to the index.  Note: its labels will be captured at this
+// time, so if the objects labels are mutated, it is important to remove the
+// item before changing its labels and then re-Add the updated item.
+//
+// To avoid the above bug, panics if the same ID is added twice.
+func (idx *LabelNameValueIndex[ItemID, Item]) Add(id ItemID, item Item) {
+	if _, ok := idx.allValues[id]; ok {
+		logrus.WithFields(logrus.Fields{
+			"id":    id,
+			"item":  item,
+			"index": idx.nameOfTrackedItems,
+		}).Panic("Add called for ID that is already in the index.")
+	}
+	idx.allValues[id] = item
+	for k, v := range item.OwnLabels() {
+		vals, ok := idx.labelNameToValueToIDs[k]
+		if !ok {
+			vals = values[ItemID]{
+				m: map[string]set.Set[ItemID]{},
+			}
+			idx.labelNameToValueToIDs[k] = vals
+		}
+		setOfIDs := vals.m[v]
+		if setOfIDs == nil {
+			setOfIDs = set.New[ItemID]()
+			vals.m[v] = setOfIDs
+		}
+		setOfIDs.Add(id)
+		vals.count++
+		idx.labelNameToValueToIDs[k] = vals
+	}
+}
+
+// Remove an item from the index.  Note that it is important that the labels
+// are not mutated between Add and Remove calls.
+func (idx *LabelNameValueIndex[ItemID, Item]) Remove(id ItemID) {
+	v := idx.allValues[id]
+	for k, v := range v.OwnLabels() {
+		vals := idx.labelNameToValueToIDs[k]
+		setOfIDs := vals.m[v]
+		setOfIDs.Discard(id)
+		if setOfIDs.Len() == 0 {
+			delete(vals.m, v)
+			if len(vals.m) == 0 {
+				delete(idx.labelNameToValueToIDs, k)
+				continue
+			}
+		}
+		vals.count--
+		idx.labelNameToValueToIDs[k] = vals
+	}
+	delete(idx.allValues, id)
+}
+
+// StrategyFor returns the best available ScanStrategy for the given
+// label name and selector LabelRestriction (which should be the restriction
+// for that label).  If the LabelRestriction is not "useful", returns
+// FullScanStrategy().
+func (idx *LabelNameValueIndex[ItemID, Item]) StrategyFor(labelName string, r parser.LabelRestriction) ScanStrategy[ItemID] {
+	if !r.MustBePresent {
+		// Not much we can do if the selector doesn't match on this label.
+		return idx.FullScanStrategy()
+	}
+
+	if r.MustHaveOneOfValues == nil {
+		// A selector such as "has(labelName)", which matches the label but
+		// not any particular value.
+		logrus.Debugf("Found %d %s with %s=<any>", idx.labelNameToValueToIDs[labelName].count, idx.nameOfTrackedItems, labelName)
```

Comment:
Good spot, I'm not sure we do.  I think we actually get the right behaviour because `idx.labelNameToValueToIDs[labelName]` will give the zero value for the `values` struct, which has a count of 0 and a nil map.  Then `EstimatedItemsToScan` will retirn 0 and Scan will iterate over the nil map (which is safe and results in no iterations).

Think I'll adjust to to avoid relying on a string of zero values.  That feels brittle.